src/server_manager/web_app/digitalocean_server.ts

```diff
@@ -38,136 +38,136 @@ const DEPRECATED_API_PORT_TAG = 'apiport';
+      // API Url and Certificate have been set, so we have successfully
+      // installed the server and can now make API calls.
+      console.info('digitalocean_server: Successfully found API and cert tags');
+      this.setInstallState(InstallState.COMPLETED);
+    } else if (tagMap.get(CERTIFICATE_FINGERPRINT_TAG)) {
+      this.setInstallState(InstallState.CERTIFICATE_CREATED);
+    } else if (tagMap.get(INSTALL_STARTED_TAG)) {
+      this.setInstallState(InstallState.DROPLET_RUNNING);
+    } else if (this.dropletInfo?.status === 'active') {
+      this.setInstallState(InstallState.DROPLET_CREATED);
     }
+  }
 
+  private isInstallTerminated(): boolean {
+    const state = this.installState.get();
+    return state === InstallState.COMPLETED ||
+        state === InstallState.ERROR ||
+        state === InstallState.DELETED;
+  }
+
+  // Maintains this.installState. Will keep polling until installation has
+  // succeeded, failed, or been canceled.
+  private async pollInstallState(): Promise<void> {
     // Periodically refresh the droplet info then try to update the install
-    // state again.
-    const intervalId = setInterval(async () => {
-      // Check if install state is already known, so we don't make an unnecessary
-      // request to fetch droplet info.
-      if (this.installState !== InstallState.UNKNOWN) {
-        clearInterval(intervalId);
-        return;
-      }
+    // state. If the final install state has been reached, don't make an
+    // unnecessary request to fetch droplet info.
+    while (!this.isInstallTerminated()) {
       try {
         await this.refreshDropletInfo();
       } catch (error) {
         console.log('Failed to get droplet info', error);
         this.setInstallState(InstallState.ERROR);
-        clearInterval(intervalId);
         return;
       }
-      updateInstallState();
-      // Immediately clear the interval if the installState is known to prevent
-      // race conditions due to setInterval firing async.
-      if (this.installState !== InstallState.UNKNOWN) {
-        clearInterval(intervalId);
+      this.updateInstallState();
+      // Return immediately if installation is terminated
+      // to prevent race conditions and avoid unnecessary delay.
+      if (this.isInstallTerminated()) {
         return;
       }
-      // Note, if there is an error refreshing the droplet, we should just
+      // TODO: If there is an error refreshing the droplet, we should just
       // try again, as there may be an intermittent network issue.
-    }, 3000);
+      await sleep(3000);
+    }
   }
 
   private setInstallState(installState: InstallState) {
-    if (this.installState !== InstallState.UNKNOWN) {
-      // Cannot change the install state once set.
-      return;
-    }
-    if (installState === InstallState.UNKNOWN) {
-      return;
+    this.installState.set(installState);
+    if (installState === InstallState.DELETED) {
+      this.installState.close();
```

Comment:
Don't you need to close whenever `isInstallTerminated()`?

Reply:
I see. That's a bit hard to follow. My suggestion is to close the stream on `isInstallTerminated`. It's ok if it gets deleted later. The goal is to monitor the installation only.